autogen/agentchat/contrib/compressible_agent.py

```diff
@@ -0,0 +1,431 @@
+            # convert trigger_count to int, default to 0.7
+            trigger_count = self.compress_config["trigger_count"]
+            if not (isinstance(trigger_count, int) or isinstance(trigger_count, float)) or trigger_count <= 0:
+                raise ValueError("trigger_count must be a positive number.")
+            if isinstance(trigger_count, float) and 0 < trigger_count <= 1:
+                self.compress_config["trigger_count"] = int(
+                    trigger_count * get_max_token_limit(self.llm_config["model"])
+                )
+            if not isinstance(self.compress_config["leave_last_n"], int) or self.compress_config["leave_last_n"] < 0:
+                raise ValueError("leave_last_n must be a non-negative integer.")
+            init_count = self._compute_init_token_count()
+            if trigger_count < init_count:
+                print(
+                    f"Warning: trigger_count {trigger_count} is less than the initial token count {init_count} (system message + function description if passed), compression will be disabled. Please increase trigger_count if you want to enable compression."
+                )
+                self.compress_config = False
+
+            if self.compress_config["mode"] == "CUSTOMIZED":
+                assert (
+                    "compress_function" in self.compress_config
+                ), "compress_function must be provided when mode is 'CUSTOMIZED'"
+            elif self.compress_config["compress_function"] is not None:
+                print("Warning: compress_function is provided but mode is not 'CUSTOMIZED'.")
+
+        else:
+            self.compress_config = False
+
+    def generate_reply(
+        self,
+        messages: Optional[List[Dict]] = None,
+        sender: Optional[Agent] = None,
+        exclude: Optional[List[Callable]] = None,
+    ) -> Union[str, Dict, None]:
+        """
+
+        Note on difference from ConversableAgent.generate_reply:
+        The following lines are removed:
+        ```
+        if messages is None:
+            messages = self._oai_messages[sender]
+        ```
+        Reason:
+        1. It will make no difference because every `generate_<>_reply` (e.g. generate_oai_reply) has the exact two lines
+            to replace `messages` with `self._oai_messages[sender]` if `messages` is None.
+
+        2. Why needed for compression:
+            - The `on_oai_token_limit` will modify `self._oai_messages[sender]`.
+            - In `generate_oai_reply`, `messages` and `sender` are passed in (line 179).
+            - If the two lines are not removed here, messages are never None, and `generate_oai_reply` will
+                always use `messages`, which is not compressed.
+            - If the two lines are removed, `generate_oai_reply` will find that `messages` is None and
+                use `self._oai_messages[sender]` instead, which is compressed.
+                (From 1, generate_oai_reply has the exact two lines)
```

Comment:
I don't understand why this is needed. `self._oai_messages[sender]` is a list. If you modify it in place in `on_oai_token_limit`, the list is modified, which means `messages` is modified too.

Reply:
Thanks for pointing this out. It turns out that removing the two lines will not help at all. The `messages` are passed in from arguments, so modifying `self._oai_messages[sender]` will not change the messages. 
I added two lines to update `messages` from  `self._oai_messages[sender]` if there is a discrepancy.